File: server/services/assure/sarif.js
/**
 * SARIF 2.1.0 exporter for Assure findings / oracle results.
 * Aligns with python/agentshield/assure/sarif.py
 */

const LEVEL = {
  Critical: 'error',
  High: 'error',
  Medium: 'warning',
  Low: 'note'
};

function toFinding(item) {
  return {
    ruleId: item.oracle_id || item.rule_id || item.name || 'finding',
    message: item.evidence || item.message || '',
    severity: item.severity || 'Medium',
    impact: item.impact || 'hygiene',
    confidence: item.confidence != null ? item.confidence : 0.5,
    taxonomy: item.taxonomy || [],
    evidence: item.evidence || '',
    passed: item.passed === true
  };
}

function findingsToSarif(items = [], options = {}) {
  const findings = items.map(toFinding).filter((f) => !f.passed);
  const rules = new Map();
  const results = [];

  for (const finding of findings) {
    if (!rules.has(finding.ruleId)) {
      rules.set(finding.ruleId, {
        id: finding.ruleId,
        name: finding.ruleId,
        shortDescription: { text: finding.message || finding.ruleId },
        fullDescription: { text: finding.evidence || finding.message || finding.ruleId },
        defaultConfiguration: { level: LEVEL[finding.severity] || 'warning' },
        properties: {
          'security-severity': finding.severity,
          impact: finding.impact,
          taxonomy: finding.taxonomy
        }
      });
    }
    results.push({
      ruleId: finding.ruleId,
      level: LEVEL[finding.severity] || 'warning',
      message: { text: finding.message || finding.ruleId },
      properties: {
        confidence: finding.confidence,
        impact: finding.impact,
        taxonomy: finding.taxonomy,
        evidence: finding.evidence
      }
    });
  }

  return {
    $schema: 'https://json.schemastore.org/sarif-2.1.0.json',
    version: '2.1.0',
    runs: [
      {
        tool: {
          driver: {
            name: options.toolName || 'agentshield',
            version: options.toolVersion || '0.1.0',
            informationUri: 'https://github.com/new-world-coder/agentshield',
            rules: [...rules.values()]
          }
        },
        results
      }
    ]
  };
}

module.exports = { findingsToSarif, toFinding };

File: python/agentshield/assure/sarif.py
"""SARIF 2.1.0 export for AgentShield findings."""

from __future__ import annotations

from typing import Any, Mapping, Sequence

from agentshield.assure.scoring import Finding, score_findings
from agentshield import __version__


_LEVEL = {
    "Critical": "error",
    "High": "error",
    "Medium": "warning",
    "Low": "note",
}


def findings_to_sarif(
    findings: Sequence[Finding | Mapping[str, Any]],
    *,
    tool_name: str = "agentshield",
    tool_version: str | None = None,
) -> dict[str, Any]:
    """Convert findings / oracle results into a SARIF 2.1.0 log."""
    report = score_findings(findings)
    rules: dict[str, dict[str, Any]] = {}
    results: list[dict[str, Any]] = []

    for finding in report.findings:
        if finding.passed:
            continue
        rule_id = finding.rule_id
        if rule_id not in rules:
            rules[rule_id] = {
                "id": rule_id,
                "name": rule_id,
                "shortDescription": {"text": finding.message or rule_id},
                "fullDescription": {
                    "text": finding.evidence or finding.message or rule_id
                },
                "defaultConfiguration": {
                    "level": _LEVEL.get(finding.severity or "Medium", "warning")
                },
                "properties": {
                    "security-severity": finding.severity,
                    "impact": finding.impact,
                    "taxonomy": finding.taxonomy,
                },
            }
        results.append(
            {
                "ruleId": rule_id,
                "level": _LEVEL.get(finding.severity or "Medium", "warning"),
                "message": {"text": finding.message or finding.evidence or rule_id},
                "properties": {
                    "confidence": finding.confidence,
                    "impact": finding.impact,
                    "taxonomy": finding.taxonomy,
                    "evidence": finding.evidence,
                    "score": report.score,
                },
            }
        )

    return {
        "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
        "version": "2.1.0",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": tool_name,
                        "version": tool_version or __version__,
                        "informationUri": "https://github.com/new-world-coder/agentshield",
                        "rules": list(rules.values()),
                    }
                },
                "results": results,
                "properties": {
                    "agentshield.score": report.score,
                    "agentshield.severity": report.severity,
                    "agentshield.statistics": report.statistics,
                },
            }
        ],
    }
